### PR TITLE
Make current xpack versions pluggable

### PR DESCRIPTION
**Pull request by thecoop**
Add pluggable extensions to `MlConfigVersion` and `TransformConfigVersion`. This is needed as we'll be exposing this version information through node info, as part of migrating away from Version entirely. It also makes it work in a similar way to `TransportVersion` and `IndexVersion`

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/VersionExtension.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * Loads a single VersionExtension, or returns {@code null} if none are found.
+     */
+    static VersionExtension load() {
+        var loader = ServiceLoader.load(VersionExtension.class);
+        var extensions = loader.stream().toList();
+        if (extensions.size() > 1) {
+            throw new IllegalStateException("More than one version extension found");
+        } else if (extensions.isEmpty()) {
+            return null;
```
Comment (ldematte) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/VersionExtension.java:43]: Nit: I see we need always to test for null when calling this, and provide a fallback.
Suggestion: return an Optional so it's clear that a VersionExtension can be missing. Or maybe, if the pattern is expected to be always to have a fallback/default, why not `VersionExtension loadOrDefault(VersionExtension)`, and pass a FallbackVersionExtension or something similar?

**Review by ldematte — APPROVED** [review]
LGTM, with one minor comment

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/VersionExtension.java (hunk 1 above)
Comment (thecoop) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/VersionExtension.java:43]: This is the same pattern as used for the other version extension handlers. I would want to change them all at once separately.

Comment (ldematte) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/VersionExtension.java:43]: Makes a lot of sense

**Comment by rjernst** [comment]
I don't think we should do this in xpack. Rather, the pluggability should be in server, where we collect and return these versions for node info.

> This is the same pattern as used for the other version extension handlers. I would want to change them all at once separately.

I plan to open a separate PR for this, it's been on my mind as I've added several of these recently.

**Comment by thecoop** [comment]
This is for xpack things, so `XPackVersionInformation`?

**Comment by rjernst** [comment]
Xpack is a lot of different modules. I don't think we should try to cram in every version concept in all of them into xpack core. Rather I think they should all plug into the extensibility in server. So IMO ML config and transforms should have their own extension classes.

**Comment by thecoop** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersionExtension.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+package org.elasticsearch.xpack.core.ml;
+
+public interface MlConfigVersionExtension {
+    /**
+     * Returns the {@link MlConfigVersion} that Elasticsearch should use.
+     * <p>
+     * This must be at least equal to the latest version found in {@link MlConfigVersion} V_* constants.
+     */
+    MlConfigVersion getCurrentMlConfigVersion();
```
Comment (rjernst) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersionExtension.java:16]: I found when actually implementing this for serverless that passing in the fallback version is important to be able to make the correct decision in the extension on which version should be used.

**Review by rjernst — APPROVED** [review]
LGTM. One suggestion.

**Review by edsavage — APPROVED** [review]
LGTM

**Comment by thecoop** [comment]
@droberts195 Is this actually something we want/need to do? Override the config versions for serverless?

**Comment by droberts195** [comment]
> Is this actually something we want/need to do?

I think it is far, far less likely that we'd ever want to introduce a new ML or transform config version in a serverless change than it is that we'll want to introduce a new transport version. We have no plans to make ML or transforms work significantly differently in serverless to how they work in stateful. So we could probably survive without this change.

However, since the PR is complete and ready to merge it seems a shame not to merge it. It means that if some unforeseen situation arises in the future where we _do_ need to introduce a new config version in serverless we have all the groundwork for that in place. So I am good to merge this. Thanks for doing all this work for us.

**Comment by thecoop** [comment]
On the other hand, I don't want to introduce complexity where none is needed - there's extra interfaces & indirection added here, which is otherwise not needed. We could leave this until we do actually need it, when we've got a pattern to easily follow here?

We're also changing how transport/indexversion works, so if we hold off here we could do a big change all at once to match those classes, if it turns out we need to in the future

**Comment by thecoop** [comment]
Closing for now - not needed at the moment

**Pull request closed without merging**